ondevice_learning.qmd

```diff
@@ -1,61 +1,630 @@
+
+To accomplish this goal, Google employed their algorithm DP-FTRL which provides a formal guarantee that trained models will not memorize specific user data or identities. DP-FTRL combined with secure aggregation, a strategy of encrypting model updates, provides an optimal balance of privacy and utility. Furthermore, adaptive clipping is applied in the aggregation process to limit the impact of individual users on the global model. Through a combination of all of these techniques, Google can continuously refine their keyboard while preserving user privacy in a formally provable way.
+
+![Google's System Design for Differential Privacy in G Board](images/ondevice_gboard_approach.png)
+
+### Benchmarking for Federated Learning: MedPerf
+
+One of the richest examples of data on the edge is medical devices. These devices store some of the most personal data on users but offer huge advances in personalized treatment and better accuracy in medical AI. Given these two factors, medical devices are the perfect use case for federated learning. [MedPerf](https://doi.org/10.1038/s42256-023-00652-2) is an open source platform used to benchmark models using federated evaluation [@karargyris2023federated]. Instead of just training models via federated learning, MedPerf takes the model to edge devices to test it against personalized data while preserving privacy. In this way a benchmark committee can evaluate various models in the real world on edge devices while still preserving patient anonymity. 
 
 ## Security Concerns
 
-Explanation: Security is a significant concern for any system that performs learning on-device, as it may expose vulnerabilities.
+Performing machine learning model training and adaptation on end-user devices also introduces security risks that must be addressed. Some key security concerns include:
+
+- **Exposure of private data**: Training data may be leaked or stolen from devices
+- **Data poisoning**: Adversaries can manipulate training data to degrade model performance
+- **Model extraction**: Attackers may attempt to steal trained model parameters
+- **Membership inference**: Models may reveal participation of specific users' data
+- **Evasion attacks**: Specially crafted inputs can cause misclassification
+
+Any system that performs learning on-device introduces security concerns, as it may expose vulnerabilities in larger scale models. There are numerous security risks associated with any machine learning model, but these risks have specific consequences for on-device learning. Fortunately, there are methods to mitigate these risks to improve the real-world performance of on-device learning. 
+
+### Data Poisoning
+
+On-device machine learning introduces unique data security challenges compared to traditional cloud-based training. In particular, data poisoning attacks pose a serious threat during on-device learning. Adversaries can manipulate training data to degrade model performance when deployed. 
```

Comment:
It might be interesting to split this into its own chapter on TinyML security/privacy, since these concerns seem more widely applicable beyond on-device learning. Frankly, data poisoning seems like less of a problem for on-device learning than elsewhere

Reply:
Yea, this is definitely something I agree with. We may just do this cleanup pass after the class.